lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -2551,3 +2567,72 @@ instance FromJSON (ApiT SmashServer) where
     parseJSON = fromTextJSON "SmashServer"
 instance ToJSON (ApiT SmashServer) where
     toJSON = toTextJSON
+
+{-------------------------------------------------------------------------------
+                         Token forging types
+-------------------------------------------------------------------------------}
+
+data ForgeTokenData (n :: NetworkDiscriminant) = ForgeTokenData
+    { forgePayments :: !(NonEmpty (AddressForgeAmount (ApiT Address, Proxy n)))
+    , assetName :: !(ApiT W.TokenName)
+    , passphrase :: !(ApiT (Passphrase "lenient"))
+    , metadata :: !(Maybe (ApiT TxMetadata))
+    , timeToLive :: !(Maybe (Quantity "second" NominalDiffTime))
+    } deriving (Eq, Generic, Show)
+
+data AddressForgeAmount addr = AddressForgeAmount
+  { address :: !addr
+  , amt :: !(Quantity "token-unit" Natural)
+  } deriving stock (Eq, Generic, Show)
+
+newtype ForgeAmount = ForgeAmount
+    { unForgeAmount
+        :: These (ApiT W.TokenMap) (ApiT W.TokenMap)
+    }
+    deriving stock (Eq, Generic)
+    deriving (Show) via Quiet ForgeAmount
+    deriving anyclass NFData
+
+data ForgeRequest
+  = Mint !Cardano.Quantity
+  | Burn !Cardano.Quantity
+
+-- next:
+-- mint: find first address in wallet and just sent tokens to that
+-- burn: say we want to send x tokens to an addr, use coin selection algorithm to find inputs, but then modify outputs to be mempty, instead burning inputs
+
+mintAmount :: ForgeAmount -> W.TokenMap
+mintAmount (ForgeAmount (This _burn))              = mempty
+mintAmount (ForgeAmount (That (ApiT mint)))        = mint
+mintAmount (ForgeAmount (These _burn (ApiT mint))) = mint
+
+burnAmount :: ForgeAmount -> W.TokenMap
+burnAmount (ForgeAmount (This (ApiT burn)))        = burn
+burnAmount (ForgeAmount (That _mint))              = mempty
+burnAmount (ForgeAmount (These (ApiT burn) _mint)) = burn
+
+instance Monoid ForgeAmount where
+  mempty = ForgeAmount $ These mempty mempty
+
+instance Semigroup ForgeAmount where
+  (ForgeAmount x) <> (ForgeAmount y) = ForgeAmount $ x <> y
```

Comment:
It might be worthwhile testing these instances with `testLaws`. For example:

https://github.com/input-output-hk/cardano-wallet/blob/1ab90cdda28e3f70a7e22e6ead568db5b98352ef/lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenMapSpec.hs#L127-L134